File: src/xuezh/core/content.py
from __future__ import annotations

from dataclasses import dataclass
import mimetypes
from pathlib import Path
import shutil
import sqlite3

from xuezh.core import clock, db, ids, paths
from xuezh.core.envelope import Artifact

ALLOWED_CONTENT_TYPES = {"story", "dialogue", "exercise"}


@dataclass(frozen=True)
class ContentResult:
    data: dict
    artifacts: list[Artifact]


def _content_destination(content_type: str, key: str, suffix: str) -> Path:
    rel = Path("cache") / "content" / content_type / f"{key}{suffix}"
    return paths.resolve_in_workspace(rel)


def _artifact_for(path: Path) -> Artifact:
    workspace = paths.ensure_workspace()
    rel_path = str(path.relative_to(workspace))
    mime, _ = mimetypes.guess_type(path.name)
    return Artifact(path=rel_path, mime=mime or "text/plain", purpose="cached_content", bytes=path.stat().st_size)


def put_content(*, content_type: str, key: str, in_path: str) -> ContentResult:
    if content_type not in ALLOWED_CONTENT_TYPES:
        raise ValueError(f"Unsupported content type: {content_type}")

    input_path = Path(in_path).expanduser()
    if not input_path.exists():
        raise FileNotFoundError(f"Input file not found: {input_path}")

    suffix = input_path.suffix or ".txt"
    dest_path = _content_destination(content_type, key, suffix)
    dest_path.parent.mkdir(parents=True, exist_ok=True)

    content_id = ids.content_id(content_type=content_type, key=key)
    db_path = db.init_db()

    conn = sqlite3.connect(db_path)
    try:
        row = conn.execute(
            "SELECT path FROM generated_content WHERE id = ?",
            (content_id,),
        ).fetchone()

        if row:
            stored_rel = Path(row[0])
            existing_path = paths.resolve_in_workspace(stored_rel)
            if not existing_path.exists():
                shutil.copy2(input_path, existing_path)
            dest_path = existing_path
        else:
            if not dest_path.exists():
                shutil.copy2(input_path, dest_path)
            rel_path_str = str(dest_path.relative_to(paths.ensure_workspace()))
            conn.execute(
                """
                INSERT INTO generated_content (id, content_type, content_key, path, created_at)
                VALUES (?, ?, ?, ?, ?)
                """,
                (content_id, content_type, key, rel_path_str, clock.now_utc().isoformat()),
            )
            conn.commit()
    finally:
        conn.close()

    artifact = _artifact_for(dest_path)
    data = {"type": content_type, "key": key, "content_id": content_id}
    return ContentResult(data=data, artifacts=[artifact])


def get_content(*, content_type: str, key: str) -> ContentResult:
    if content_type not in ALLOWED_CONTENT_TYPES:
        raise ValueError(f"Unsupported content type: {content_type}")

    content_id = ids.content_id(content_type=content_type, key=key)
    db_path = db.init_db()

    conn = sqlite3.connect(db_path)
    try:
        row = conn.execute(
            "SELECT path FROM generated_content WHERE id = ?",
            (content_id,),
        ).fetchone()
        if not row:
            raise FileNotFoundError(f"Content not found for key: {key}")
        rel_path = Path(row[0])
        resolved = paths.resolve_in_workspace(rel_path)
        if not resolved.exists():
            raise FileNotFoundError(f"Cached content missing on disk: {resolved}")
    finally:
        conn.close()

    artifact = _artifact_for(resolved)
    data = {"type": content_type, "key": key, "content_id": content_id}
    return ContentResult(data=data, artifacts=[artifact])

File: internal/xuezh/content/content.go
package content

import (
	"database/sql"
	"errors"
	"fmt"
	"mime"
	"os"
	"path/filepath"
	"strings"

	_ "github.com/mattn/go-sqlite3"

	"github.com/joshp123/xuezh/internal/xuezh/clock"
	"github.com/joshp123/xuezh/internal/xuezh/db"
	"github.com/joshp123/xuezh/internal/xuezh/envelope"
	"github.com/joshp123/xuezh/internal/xuezh/ids"
	"github.com/joshp123/xuezh/internal/xuezh/paths"
)

var allowedContentTypes = map[string]struct{}{
	"story":    {},
	"dialogue": {},
	"exercise": {},
}

type ContentResult struct {
	Data      map[string]any
	Artifacts []envelope.Artifact
}

func contentDestination(contentType, key, suffix string) (string, error) {
	rel := filepath.Join("cache", "content", contentType, key+suffix)
	return paths.ResolveInWorkspace(rel)
}

func artifactFor(path string) (envelope.Artifact, error) {
	workspace, err := paths.EnsureWorkspace()
	if err != nil {
		return envelope.Artifact{}, err
	}
	rel, err := relativeTo(workspace, path)
	if err != nil {
		return envelope.Artifact{}, err
	}
	mimeType := mime.TypeByExtension(filepath.Ext(path))
	if mimeType == "" {
		mimeType = "text/plain"
	}
	if idx := strings.Index(mimeType, ";"); idx != -1 {
		mimeType = strings.TrimSpace(mimeType[:idx])
	}
	stat, err := os.Stat(path)
	if err != nil {
		return envelope.Artifact{}, err
	}
	return envelope.Artifact{Path: rel, MIME: mimeType, Purpose: "cached_content", Bytes: intPtr(int(stat.Size()))}, nil
}

func PutContentBytes(contentType, key, filename string, data []byte) (ContentResult, error) {
	if _, ok := allowedContentTypes[contentType]; !ok {
		return ContentResult{}, errors.New("Unsupported content type: " + contentType)
	}
	suffix := filepath.Ext(filename)
	if suffix == "" {
		suffix = ".txt"
	}
	destPath, err := contentDestination(contentType, key, suffix)
	if err != nil {
		return ContentResult{}, err
	}
	if err := os.MkdirAll(filepath.Dir(destPath), 0o755); err != nil {
		return ContentResult{}, err
	}
	contentID := ids.ContentID(contentType, key)
	dbPath, err := db.InitDB()
	if err != nil {
		return ContentResult{}, err
	}
	conn, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return ContentResult{}, err
	}
	defer conn.Close()
	var storedRel string
	row := conn.QueryRow("SELECT path FROM generated_content WHERE id = ?", contentID)
	switch err := row.Scan(&storedRel); err {
	case nil:
		resolved, err := paths.ResolveInWorkspace(storedRel)
		if err != nil {
			return ContentResult{}, err
		}
		if _, err := os.Stat(resolved); os.IsNotExist(err) {
			if err := writeFile(resolved, data); err != nil {
				return ContentResult{}, err
			}
		}
		destPath = resolved
	case sql.ErrNoRows:
		if _, err := os.Stat(destPath); os.IsNotExist(err) {
			if err := writeFile(destPath, data); err != nil {
				return ContentResult{}, err
			}
		}
		workspace, err := paths.EnsureWorkspace()
		if err != nil {
			return ContentResult{}, err
		}
		relPath, err := relativeTo(workspace, destPath)
		if err != nil {
			return ContentResult{}, err
		}
		now, err := clock.NowUTC()
		if err != nil {
			return ContentResult{}, err
		}
		_, err = conn.Exec(
			`INSERT INTO generated_content (id, content_type, content_key, path, created_at)
			 VALUES (?, ?, ?, ?, ?)`,
			contentID, contentType, key, relPath, clock.FormatISO(now),
		)
		if err != nil {
			return ContentResult{}, err
		}
	default:
		return ContentResult{}, err
	}
	artifact, err := artifactFor(destPath)
	if err != nil {
		return ContentResult{}, err
	}
	payload := map[string]any{"type": contentType, "key": key, "content_id": contentID}
	return ContentResult{Data: payload, Artifacts: []envelope.Artifact{artifact}}, nil
}

func GetContent(contentType, key string) (ContentResult, error) {
	if _, ok := allowedContentTypes[contentType]; !ok {
		return ContentResult{}, errors.New("Unsupported content type: " + contentType)
	}
	contentID := ids.ContentID(contentType, key)
	dbPath, err := db.InitDB()
	if err != nil {
		return ContentResult{}, err
	}
	conn, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return ContentResult{}, err
	}
	defer conn.Close()
	var storedRel string
	row := conn.QueryRow("SELECT path FROM generated_content WHERE id = ?", contentID)
	if err := row.Scan(&storedRel); err != nil {
		if err == sql.ErrNoRows {
			return ContentResult{}, errors.New("Content not found for key: " + key)
		}
		return ContentResult{}, err
	}
	resolved, err := paths.ResolveInWorkspace(storedRel)
	if err != nil {
		return ContentResult{}, err
	}
	if _, err := os.Stat(resolved); err != nil {
		return ContentResult{}, errors.New("Cached content missing on disk: " + resolved)
	}
	artifact, err := artifactFor(resolved)
	if err != nil {
		return ContentResult{}, err
	}
	data := map[string]any{"type": contentType, "key": key, "content_id": contentID}
	return ContentResult{Data: data, Artifacts: []envelope.Artifact{artifact}}, nil
}

func GetContentBytes(contentType, key string) (ContentResult, []byte, error) {
	result, err := GetContent(contentType, key)
	if err != nil {
		return ContentResult{}, nil, err
	}
	if len(result.Artifacts) == 0 {
		return ContentResult{}, nil, errors.New("Cached content has no artifact")
	}
	path, err := paths.ResolveInWorkspace(result.Artifacts[0].Path)
	if err != nil {
		return ContentResult{}, nil, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return ContentResult{}, nil, err
	}
	return result, data, nil
}

func writeFile(path string, data []byte) error {
	out, err := os.Create(path)
	if err != nil {
		return err
	}
	defer func() {
		_ = out.Close()
	}()
	if _, err := out.Write(data); err != nil {
		return err
	}
	return out.Sync()
}

func relativeTo(base, target string) (string, error) {
	baseClean := cleanExistingPath(base)
	targetClean := cleanExistingPath(target)
	if targetClean != baseClean && !strings.HasPrefix(targetClean, baseClean+string(filepath.Separator)) {
		return "", fmt.Errorf("'%s' is not in the subpath of '%s'", target, base)
	}
	return filepath.Rel(baseClean, targetClean)
}

func cleanExistingPath(path string) string {
	abs, err := filepath.Abs(path)
	if err == nil {
		path = abs
	}
	resolved, err := filepath.EvalSymlinks(path)
	if err == nil {
		path = resolved
	}
	return filepath.Clean(path)
}

func intPtr(value int) *int {
	return &value
}
